### PR TITLE
Sentry integration

### LINKED ISSUE #217: Sentry error logging
Once we get this into production, we'll want sentry error logs. Sentry supports both Python and JS environments, so we should be able to get full-stack logging. This logging should be optional, and shouldn't happen in development environments.

Sentry allows for a lot of metadata, so try to hook it up with auth'd users wherever possible.

More information about sentry over at https://sentry.io/

### PR DESCRIPTION
**Pull request by AMStrix**
Closes #217.

### What this does
- sentry for BE, FE (browser & server)
- `user` scope for BE and FE browser
- `release` defaults to short git HEAD hash
- other than setting the user scope, this is just the basics!

### Notes
- we will want to integrate with github once we have the company Sentry account
- currently unhandled errors are gobbled up by YOLO -- thus Sentry will not report them

### Test
- install new deps for FE and BE
- get a sentry account and set `SENTRY_DSN` in both `frontend` & `backend`
- also set `SENTRY_ENVIRONMENT` to 'development'

#### backend
- temporarily create and catch an error in a `requires_sm` `user/views.py:128 (update_user)` 
```
from sentry_sdk import capture_exception
...
    # inside update_user
    try:
        raise Exception('Python error test.')
    except Exception as e:
        capture_exception(e)
```
1. trigger the route by editing the user profile through the app
1. affirm the error appears in Sentry and that user `email`, `accountAddress`, `ip_address` & `id` is populated.
1. affirm that `release` & `environment` are also populated
screen shot 2018-11-19 at 2 54 58 pm
screen shot 2018-11-19 at 3 33 41 pm

#### frontend (browser)
- temporarily add error to `client/index.tsx:25`
```
window.setTimeout(() => {
  throw new Error('Browser error test.');
}, 5000);
```
1. load a page as an authenticated user, wait 5 seconds
1. affirm the error appears in Sentry and that user `email` & `accountAddress` is populated.
1. affirm that `release` & `environment` are also populated

#### frontend (server)
- temporarily add error to `server/index.tsx:33`
```
app.get('/', function mainHandler(req, res) {
  throw new Error('Node error test.');
});
```
1. load the page `http://localhost:3000/`, see nasty error
1. affirm the error appears in Sentry
1. affirm that `release` & `environment` are populated

### CODE REVIEW

File: backend/.env.example [hunk 1]
```diff
@@ -7,5 +7,8 @@ REDISTOGO_URL="redis://localhost:6379"
...
 SENDGRID_API_KEY="optional, but emails won't send without it"
 # for ropsten use the following
-# ETHEREUM_ENDPOINT_URI = "https://ropsten.infura.io/API_KEY"
-ETHEREUM_ENDPOINT_URI = "http://localhost:8545"
\ No newline at end of file
+# ETHEREUM_ENDPOINT_URI="https://ropsten.infura.io/API_KEY"
+ETHEREUM_ENDPOINT_URI="http://localhost:8545"
+SENTRY_DSN="https://PUBLICKEY@sentry.io/PROJECTID"
+SENTRY_RELEASE="optional, overrides git hash"
+SENTRY_ENVIRONMENT=development
```
Comment (wbobeirne) [inline backend/.env.example]: We should probably just piggy-back off of FLASK_ENV for this

File: backend/grant/utils/auth.py [hunk 2]
```diff
@@ -69,6 +70,13 @@ def decorated(*args, **kwargs):
                 return jsonify(message="No user exists with address: {}".format(auth_address)), 401
 
             g.current_user = user
+            with sentry_sdk.configure_scope() as scope:
+                scope.user = {
+                    "id": user.id,
+                    "email": user.email_address,
+                    "accountAddress": user.account_address,
+                    "ip_address": request.remote_addr,
+                }
```
Comment (wbobeirne) [inline backend/grant/utils/auth.py]: Perhaps we should keep it to ID & IP, and do a lookup for extra info on our end. I'd rather not have to put in our privacy policy that Sentry gets the whole enchilada. Maybe even just ID, not sure if I feel that IP buys us much other than maybe naive spam detection.

File: frontend/.envexample [hunk 3]
```diff
@@ -5,4 +5,9 @@ FUND_ETH_ADDRESSES=0x4bbeEB066eD09B7AEd07bF39EEe0460DFa261520,0xDECAF9CD2367cdbb
...
 
 # Set the public host url (no trailing slash)
-PUBLIC_HOST_URL=https://demo.grant.io
\ No newline at end of file
+PUBLIC_HOST_URL=https://demo.grant.io
+
+# sentry 
+SENTRY_DSN="https://PUBLICKEY@sentry.io/PROJECTID"
+SENTRY_RELEASE="optional, overrides git hash"
+SENTRY_ENVIRONMENT=development
```
Comment (wbobeirne) [inline frontend/.envexample]: We should just use `process.env.NODE_ENV`.

**Comment by AMStrix** [comment]
FYI - fixed issues.

**Comment by dternyak** [comment]
Going off of Will's approval here since I don't have sentry creds!

**Pull request merged**